File: Reaktoro/Common/Table.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2022 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *
import pytest


def testTable():

    # =======================================================================================================
    # Testing basic usage of Table and TableColumn
    # =======================================================================================================
    table = Table()

    #----------------------------------------------------------------------------------------------------
    # Checking methods TableColumn.appendFloat|appendInteger|appendString|appendBoolean
    #----------------------------------------------------------------------------------------------------

    table.column("Floats").appendFloat(0.0)
    table.column("Floats").appendFloat(1.0)
    table.column("Floats").appendFloat(2.0)
    table.column("Floats").appendFloat(3.0)
    table.column("Floats").appendFloat(4.0)

    table.column("Integers").appendInteger(2)
    table.column("Integers").appendInteger(3)
    table.column("Integers").appendInteger(4)
    table.column("Integers").appendInteger(5)

    table.column("Strings").appendString("Hello")
    table.column("Strings").appendString("World")
    table.column("Strings").appendString("!")

    table.column("Booleans").appendBoolean(True)
    table.column("Booleans").appendBoolean(False)

    #----------------------------------------------------------------------------------------------------
    # Checking methods Table.rows and Table.cols
    #----------------------------------------------------------------------------------------------------

    assert table.rows() == 5  # the number of rows in column Floats
    assert table.cols() == 4  # the number of columns registered above

    #----------------------------------------------------------------------------------------------------
    # Checking method TableColumn.rows
    #----------------------------------------------------------------------------------------------------

    assert table.column("Floats").rows()   == 5
    assert table.column("Integers").rows() == 4
    assert table.column("Strings").rows()  == 3
    assert table.column("Booleans").rows() == 2

    #----------------------------------------------------------------------------------------------------
    # Checking methods TableColumn.dataType
    #----------------------------------------------------------------------------------------------------

    assert table.column("Floats").dataType()   == TableColumn.DataType.Float
    assert table.column("Integers").dataType() == TableColumn.DataType.Integer
    assert table.column("Strings").dataType()  == TableColumn.DataType.String
    assert table.column("Booleans").dataType() == TableColumn.DataType.Boolean

    #----------------------------------------------------------------------------------------------------
    # Checking conversion methods TableColumn.floats|integers|strings|booleans
    #----------------------------------------------------------------------------------------------------

    assert table.column("Floats").floats()     == [0.0, 1.0, 2.0, 3.0, 4.0]
    assert table.column("Integers").integers() == [2, 3, 4, 5]
    assert table.column("Strings").strings()   == ["Hello", "World", "!"]
    assert table.column("Booleans").booleans() == [True, False]

    with pytest.raises(Exception): table.column("Floats").integers()
    with pytest.raises(Exception): table.column("Floats").strings()
    with pytest.raises(Exception): table.column("Floats").booleans()

    with pytest.raises(Exception): table.column("Integers").floats()
    with pytest.raises(Exception): table.column("Integers").strings()
    with pytest.raises(Exception): table.column("Integers").booleans()

    with pytest.raises(Exception): table.column("Strings").floats()
    with pytest.raises(Exception): table.column("Strings").integers()
    with pytest.raises(Exception): table.column("Strings").booleans()

    with pytest.raises(Exception): table.column("Booleans").floats()
    with pytest.raises(Exception): table.column("Booleans").integers()
    with pytest.raises(Exception): table.column("Booleans").strings()

    #----------------------------------------------------------------------------------------------------
    # Checking method TableColumn.append
    #----------------------------------------------------------------------------------------------------

    try: table.column("Floats").append(5)  # integer 5 can be inserted with append in a column of floats!
    except Exception: pytest.fail("An exception was not expected here.")

    try: table.column("Floats").append(6.0)  # float 5.0 can be inserted with append in a column of floats - no problem!
    except Exception: pytest.fail("An exception was not expected here.")

    with pytest.raises(Exception): table.column("Floats").appendInteger(5)  # integer 5 cannot be inserted with appendInteger in a column of floats!
    with pytest.raises(Exception): table.column("Floats").append("NOT OK")
    with pytest.raises(Exception): table.column("Floats").append(True)

    try: table.column("Integers").append(6)  # an integer can be appended to a column of integers - no problem!
    except Exception: pytest.fail("An exception was not expected here.")

    with pytest.raises(Exception): table.column("Integers").append(1.0)  # a float cannot be appended to a column of integers!
    with pytest.raises(Exception): table.column("Integers").append("NOT OK")
    with pytest.raises(Exception): table.column("Integers").append(False)

    try: table.column("Strings").append("Star")  # a string can be appended to a column of strings - no problem!
    except Exception: pytest.fail("An exception was not expected here.")

    with pytest.raises(Exception): table.column("Strings").append(1.0)
    with pytest.raises(Exception): table.column("Strings").append(3)
    with pytest.raises(Exception): table.column("Strings").append(False)

    try: table.column("Booleans").append(True)  # a boolean can be appended to a column of booleans - no problem!
    except Exception: pytest.fail("An exception was not expected here.")

    with pytest.raises(Exception): table.column("Booleans").append(0.0)
    with pytest.raises(Exception): table.column("Booleans").append(8)
    with pytest.raises(Exception): table.column("Booleans").append("False")

    #----------------------------------------------------------------------------------------------------
    # Checking operator Table.operator[]
    #----------------------------------------------------------------------------------------------------

    # Check it on column of floats
    assert table["Floats"][0] == 0.0
    assert table["Floats"][1] == 1.0
    assert table["Floats"][2] == 2.0
    assert table["Floats"][3] == 3.0
    assert table["Floats"][4] == 4.0

    # Check it on columns of non-floats (Table.operator[] only works with column of floats!)
    with pytest.raises(Exception): table["Integers"]
    with pytest.raises(Exception): table["Strings"]
    with pytest.raises(Exception): table["Booleans"]

    #----------------------------------------------------------------------------------------------------
    # Checking method Table.dump
    #----------------------------------------------------------------------------------------------------

    assert table.dump() == (
        " Floats | Integers | Strings | Booleans\n"
        "0.00000 |        2 |   Hello |        1\n"
        "1.00000 |        3 |   World |        0\n"
        "2.00000 |        4 |       ! |        1\n"
        "3.00000 |        5 |    Star |         \n"
        "4.00000 |        6 |         |         \n"
        "5.00000 |          |         |         ")

    table.column("Column name with space").append(1.0)
    table.column("Column name with space").append(2.0)

    assert table.dump() == (
        " Floats | Integers | Strings | Booleans | Column name with space\n"
        "0.00000 |        2 |   Hello |        1 |                1.00000\n"
        "1.00000 |        3 |   World |        0 |                2.00000\n"
        "2.00000 |        4 |       ! |        1 |                       \n"
        "3.00000 |        5 |    Star |          |                       \n"
        "4.00000 |        6 |         |          |                       \n"
        "5.00000 |          |         |          |                       ")

    opts = Table.OutputOptions()
    opts.delimiter = " | "
    opts.precision = 4
    opts.scientific = True

    assert table.dump(opts) == (
        "    Floats | Integers | Strings | Booleans | Column name with space\n"
        "0.0000e+00 |        2 |   Hello |        1 |             1.0000e+00\n"
        "1.0000e+00 |        3 |   World |        0 |             2.0000e+00\n"
        "2.0000e+00 |        4 |       ! |        1 |                       \n"
        "3.0000e+00 |        5 |    Star |          |                       \n"
        "4.0000e+00 |        6 |         |          |                       \n"
        "5.0000e+00 |          |         |          |                       ")

    # =======================================================================================================
    # Testing a more convenient usage of Table using operator<<
    # =======================================================================================================
    table = Table()

    #----------------------------------------------------------------------------------------------------
    # Checking operator TableColumn.operator<<
    #----------------------------------------------------------------------------------------------------

    table.column("Floats") << 10.0 << 20 << 30.0 << 40 << 50.0 << 60 << 70.0  # integers 20, 40, 60 will be cast to floats
    table.column("Integers") << 1 << 2 << 3 << 4 << 5
    table.column("Strings") << "Hello" << "World" << "!"
    table.column("Booleans") << True << False

    assert table.rows() == 7
    assert table.cols() == 4

    assert table.column("Floats").rows()   == 7
    assert table.column("Integers").rows() == 5
    assert table.column("Strings").rows()  == 3
    assert table.column("Booleans").rows() == 2

    assert table.column("Floats").dataType()   == TableColumn.DataType.Float
    assert table.column("Integers").dataType() == TableColumn.DataType.Integer
    assert table.column("Strings").dataType()  == TableColumn.DataType.String
    assert table.column("Booleans").dataType() == TableColumn.DataType.Boolean

    assert table.column("Floats").floats()     == [10.0, 20.0, 30.0, 40.0, 50.0, 60.0, 70.0]
    assert table.column("Integers").integers() == [1, 2, 3, 4, 5]
    assert table.column("Strings").strings()   == ["Hello", "World", "!"]
    assert table.column("Booleans").booleans() == [True, False]

    #----------------------------------------------------------------------------------------------------
    # Checking method TableColumn.dump
    #----------------------------------------------------------------------------------------------------

    assert table.dump() == (
        " Floats | Integers | Strings | Booleans\n"
        "10.0000 |        1 |   Hello |        1\n"
        "20.0000 |        2 |   World |        0\n"
        "30.0000 |        3 |       ! |         \n"
        "40.0000 |        4 |         |         \n"
        "50.0000 |        5 |         |         \n"
        "60.0000 |          |         |         ")

File: Reaktoro/Common/Table.cpp
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2022 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

#include "Table.hpp"

// C++ includes
#include <sstream>
#include <fstream>
#include <iomanip>

// Reaktoro includes
#include <Reaktoro/Common/Algorithms.hpp>
#include <Reaktoro/Common/Constants.hpp>
#include <Reaktoro/Common/Enumerate.hpp>

namespace Reaktoro {
namespace {

// Convenient alias for this translation unit.
using DataType = TableColumn::DataType;

/// Return true if type `T` is consistent with given column data type.
template<typename T>
auto isTypeConsistent(DataType datatype) -> bool
{
    switch(datatype)
    {
        case DataType::Float: return isFloatingPoint<T> || isInteger<T>;
        case DataType::Integer: return isInteger<T>;
        case DataType::String: return isSame<T, String>;
        case DataType::Boolean: return isSame<T, bool>;
        case DataType::Undefined: return false;
        default: return false;
    }
}

/// Return a string representation for a DataType value.
auto strColumnDataType(DataType datatype) -> String
{
    switch(datatype)
    {
        case DataType::Float: return "floating-point";
        case DataType::Integer: return "integer";
        case DataType::String: return "string";
        case DataType::Boolean: return "boolean";
        default: return "Undefined";
    }
}

/// General implementation of an append function to add a new value to a table column.
template<DataType type, typename T>
auto appendNewValue(Any& data, DataType& datatype, T const& value, Chars strvaluetype)
{
    if(!data.has_value())
    {
        data = Deque<T>();
        datatype = type;
    }

    errorifnot(isTypeConsistent<T>(type),
        "You cannot append a value of ", strvaluetype, " type to a table column that store values of ", strColumnDataType(type), " type. "
        "Make sure that after inserting the first value into a table column, the same value type is used for all subsequent inserts. "
        "Note that integer values can be stored as floating-point values in a table column of floats. No other conversion is supported.");

    auto& deque = std::any_cast<Deque<T>&>(data);
    deque.push_back(value);
}

/// Convert a TableColumn object to a vector of strings representing the column's data along its rows.
auto stringfyTableColumn(String colname, TableColumn const& column, Table::OutputOptions const& outputopts) -> Strings
{
    auto const& [delimiter, precision, scientific] = outputopts;

    Strings rows;
    rows.reserve(1 + column.rows()); // extra entry for column's name (first entry!)

    auto columnname_has_delimiter = colname.find(delimiter) != String::npos;
    rows.push_back(columnname_has_delimiter ? "\"" + colname + "\"" : colname); // e.g., delimiter is `,` and column name is `alpha,beta`; name becomes "alpha,beta" in the output

    std::ostringstream ss;
    ss << (scientific ? std::scientific : std::fixed);
    ss << std::showpoint;
    ss << std::setprecision(precision);

    for(auto i = 0; i < column.rows(); ++i)
    {
        ss.str("");
        std::visit([&](auto arg){ ss << arg; }, column[i]);
        rows.push_back(ss.str());
    }
    return rows;
}

/// Compute the maximum width of a table column already stringfied.
auto widthTableColumn(Strings const& rows) -> Index
{
    Index width = 0;
    for(auto const& row : rows)
        width = std::max(width, row.size());
    return width;
}

/// Assemble a vector with the width of each column in a table already stringfied.
auto widthsTableColumns(Vec<Strings> const& matrix) -> Vec<Index>
{
    return vectorize(matrix, RKT_LAMBDA(x, widthTableColumn(x)));
}

/// Output a Table object to a general stream object with given delimiter and numeric precision.
template<typename Stream>
auto outputTable(Stream& stream, Table const& table, Table::OutputOptions const& outputopts)
{
    // Stringfy all columns, creating a matrix of strings
    const Vec<Strings> matrix = vectorize(table.columns(), RKT_LAMBDA(pair, stringfyTableColumn(pair.first, pair.second, outputopts)));

    // Compute the maximum widths in each table column for improved output format with alignment
    const Vec<Index> widths = widthsTableColumns(matrix);

    // Define an auxiliary function to determine if a delimiter symbol is to be inserted or not
    const auto space = String(" ");
    const auto delim = [&](auto j) { return j > 0 ? outputopts.delimiter + space : space; };

    // Output the table, row by row
    for(auto i = 0; i < table.rows(); ++i) {
        for(auto j = 0; j < table.cols(); ++j)
            stream << delim(j) << std::setw(widths[j]) << (i < matrix[j].size() ? matrix[j][i] : ""); // matrix[j] is the j-th column, and matrix[j][i] is the i-th entry in the j-th column
        stream << "\n";
    }
}

} // anonymous namespace

TableColumn::TableColumn()
{}

auto TableColumn::appendFloat(double value) -> void
{
    appendNewValue<DataType::Float>(data, datatype, value, "floating-point");
    ++mrows;
}

auto TableColumn::appendInteger(long value) -> void
{
    appendNewValue<DataType::Integer>(data, datatype, value, "integer");
    ++mrows;
}

auto TableColumn::appendString(String const& value) -> void
{
    appendNewValue<DataType::String>(data, datatype, value, "string");
    ++mrows;
}

auto TableColumn::appendBoolean(bool value) -> void
{
    appendNewValue<DataType::Boolean>(data, datatype, value, "boolean");
    ++mrows;
}

auto TableColumn::dataType() const -> DataType
{
    return datatype;
}

auto TableColumn::floats() const -> Deque<double> const&
{
    errorif(datatype != DataType::Float, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to values a list of of floating-point type.");
    return std::any_cast<Deque<double> const&>(data);
}

auto TableColumn::floats() -> Deque<double>&
{
    errorif(datatype != DataType::Float, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to values a list of of floating-point type.");
    return std::any_cast<Deque<double>&>(data);
}

auto TableColumn::integers() const -> Deque<long> const&
{
    errorif(datatype != DataType::Integer, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of integer type.");
    return std::any_cast<Deque<long> const&>(data);
}

auto TableColumn::integers() -> Deque<long>&
{
    errorif(datatype != DataType::Integer, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of integer type.");
    return std::any_cast<Deque<long>&>(data);
}

auto TableColumn::strings() const -> Deque<String> const&
{
    errorif(datatype != DataType::String, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of string type.");
    return std::any_cast<Deque<String> const&>(data);
}

auto TableColumn::strings() -> Deque<String>&
{
    errorif(datatype != DataType::String, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of string type.");
    return std::any_cast<Deque<String>&>(data);
}

auto TableColumn::booleans() const -> Deque<bool> const&
{
    errorif(datatype != DataType::Boolean, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of boolean type.");
    return std::any_cast<Deque<bool> const&>(data);
}

auto TableColumn::booleans() -> Deque<bool>&
{
    errorif(datatype != DataType::Boolean, "You cannot convert a table column with values of ", strColumnDataType(datatype), " type to a list of values of boolean type.");
    return std::any_cast<Deque<bool>&>(data);
}

auto TableColumn::rows() const -> Index
{
    return mrows;
}

auto TableColumn::operator[](Index row) const -> std::variant<double, long, String, bool>
{
    errorifnot(row < mrows, "Given row index, ", row, ", is greater than number of rows in the table column, ", mrows, ".");
    switch(datatype)
    {
        case DataType::Float:     return std::any_cast<Deque<double> const&>(data)[row];
        case DataType::Integer:   return std::any_cast<Deque<long> const&>(data)[row];
        case DataType::String:    return std::any_cast<Deque<String> const&>(data)[row];
        case DataType::Boolean:   return std::any_cast<Deque<bool> const&>(data)[row];
        default: return NaN;
    }
}

Table::Table()
{

}

auto Table::columns() const -> Dict<String, TableColumn> const&
{
    return mcolumns;
}

auto Table::column(String const& columnname) const -> TableColumn const&
{
    auto it = mcolumns.find(columnname);
    errorif(it == mcolumns.end(), "There is no column named `", columnname, "` in Table object.");
    return it->second;
}

auto Table::column(String const& columnname) -> TableColumn&
{
    return mcolumns[columnname];
}

auto Table::operator[](String const& columnname) const -> Deque<double> const&
{
    return std::as_const(*this)[columnname];
}

auto Table::operator[](String const& columnname) -> Deque<double>&
{
    auto& col = column(columnname);
    auto const& datatype = col.dataType();
    errorif(datatype != DataType::Float,
        "You are using operator [] on a Table object to retrieve float values of a column that stores ", strColumnDataType(datatype), " values. "
        "Use one of the cast methods in class Table to achieve your desired result.");
    return col.floats();
}

auto Table::rows() const -> Index
{
    Index size = 0;
    for(auto const& [name, column]: mcolumns)
        size = std::max(size, column.rows());
    return size;
}

auto Table::cols() const -> Index
{
    return mcolumns.size();
}

auto Table::dump(OutputOptions const& outputopts) const -> String
{
    std::stringstream ss;
    outputTable(ss, *this, outputopts);
    return ss.str();
}

auto Table::save(String const& filepath, OutputOptions const& outputopts) const -> void
{
    std::ofstream file(filepath);
    outputTable(file, *this, outputopts);
}

Table::OutputOptions::OutputOptions()
: delimiter(""), precision(6), scientific(false)
{}

} // namespace Reaktoro
